File: app/api/_hospital_catalog.ts
export type HospitalDepartment = {
  code: string;
  name: string;
  aliases: string[];
  symptoms: string[];
  doctors: string[];
};

type DepartmentSeed = Omit<HospitalDepartment, "doctors">;

const DEPARTMENT_SEEDS: DepartmentSeed[] = [
  { code: "general-medicine", name: "General Medicine (Internal Medicine)", aliases: ["general medicine", "internal medicine", "internist"], symptoms: ["persistent fever", "fatigue", "unexplained weight loss", "multiple symptoms", "adult checkup"] },
  { code: "family-medicine", name: "Family Medicine", aliases: ["family medicine", "family doctor", "primary care"], symptoms: ["routine checkup", "cough", "minor illness", "preventive care", "ongoing primary care"] },
  { code: "emergency-medicine", name: "Emergency Medicine", aliases: ["emergency medicine", "emergency department", "emergency room"], symptoms: ["severe injury", "chest pain", "difficulty breathing", "loss of consciousness", "severe bleeding"] },
  { code: "cardiology", name: "Cardiology", aliases: ["cardiology", "cardiologist", "cardiac clinic"], symptoms: ["palpitations", "chest discomfort", "high blood pressure", "leg swelling", "abnormal ecg"] },
  { code: "neurology", name: "Neurology", aliases: ["neurology", "neurologist"], symptoms: ["headache", "seizure", "numbness", "tremor", "memory problems"] },
  { code: "neurosurgery", name: "Neurosurgery", aliases: ["neurosurgery", "neurosurgeon"], symptoms: ["brain surgery consultation", "spine surgery consultation", "nerve compression", "brain tumor surgery", "spinal cord compression"] },
  { code: "orthopedic-surgery", name: "Orthopedic Surgery", aliases: ["orthopedic surgery", "orthopaedic surgery", "orthopedics", "orthopaedics"], symptoms: ["joint pain", "leg pain", "fracture", "back pain", "sports injury"] },
  { code: "general-surgery", name: "General Surgery", aliases: ["general surgery", "general surgeon"], symptoms: ["hernia", "gallbladder problem", "appendix problem", "abdominal lump", "surgical wound"] },
  { code: "plastic-surgery", name: "Plastic Surgery", aliases: ["plastic surgery", "plastic surgeon", "reconstructive surgery"], symptoms: ["burn reconstruction", "scar revision", "hand reconstruction", "complex wound", "facial reconstruction"] },
  { code: "vascular-surgery", name: "Vascular Surgery", aliases: ["vascular surgery", "vascular surgeon"], symptoms: ["varicose veins", "poor circulation", "leg ulcer", "arterial blockage", "carotid disease"] },
  { code: "thoracic-surgery", name: "Thoracic Surgery", aliases: ["thoracic surgery", "thoracic surgeon", "chest surgery"], symptoms: ["lung surgery consultation", "chest wall mass", "mediastinal mass", "esophageal surgery", "pleural disease"] },
  { code: "cardiac-surgery", name: "Cardiac Surgery", aliases: ["cardiac surgery", "heart surgery", "cardiothoracic surgery"], symptoms: ["bypass surgery consultation", "heart valve surgery", "aortic surgery", "congenital heart surgery", "post cardiac surgery review"] },
  { code: "gastroenterology", name: "Gastroenterology", aliases: ["gastroenterology", "gastroenterologist", "digestive clinic"], symptoms: ["abdominal pain", "acid reflux", "diarrhea", "constipation", "difficulty swallowing"] },
  { code: "hepatology", name: "Hepatology", aliases: ["hepatology", "hepatologist", "liver clinic"], symptoms: ["jaundice", "hepatitis", "fatty liver", "cirrhosis", "abnormal liver tests"] },
  { code: "pulmonology", name: "Pulmonology", aliases: ["pulmonology", "pulmonologist", "respiratory medicine", "lung clinic"], symptoms: ["chronic cough", "wheezing", "shortness of breath", "sleep apnea", "abnormal lung test"] },
  { code: "nephrology", name: "Nephrology", aliases: ["nephrology", "nephrologist", "kidney clinic"], symptoms: ["chronic kidney disease", "abnormal creatinine", "protein in urine", "dialysis care", "electrolyte problem"] },
  { code: "urology", name: "Urology", aliases: ["urology", "urologist"], symptoms: ["urinary difficulty", "kidney stones", "prostate problem", "blood in urine", "bladder problem"] },
  { code: "endocrinology", name: "Endocrinology", aliases: ["endocrinology", "endocrinologist", "hormone clinic"], symptoms: ["diabetes", "thyroid problem", "hormone disorder", "adrenal problem", "metabolic disorder"] },
  { code: "rheumatology", name: "Rheumatology", aliases: ["rheumatology", "rheumatologist"], symptoms: ["joint stiffness", "autoimmune disease", "lupus", "inflammatory arthritis", "vasculitis"] },
  { code: "hematology", name: "Hematology", aliases: ["hematology", "haematology", "hematologist"], symptoms: ["anemia", "bleeding disorder", "clotting disorder", "abnormal blood count", "bone marrow problem"] },
  { code: "oncology", name: "Oncology", aliases: ["oncology", "oncologist", "cancer clinic"], symptoms: ["cancer evaluation", "chemotherapy consultation", "tumor follow-up", "cancer survivorship", "new cancer diagnosis"] },
  { code: "dermatology", name: "Dermatology", aliases: ["dermatology", "dermatologist", "skin clinic"], symptoms: ["skin rash", "acne", "changing mole", "skin lesion", "eczema"] },
  { code: "ophthalmology", name: "Ophthalmology", aliases: ["ophthalmology", "ophthalmologist", "eye clinic"], symptoms: ["vision changes", "eye pain", "cataract", "glaucoma", "retina problem"] },
  { code: "ent", name: "ENT (Otolaryngology)", aliases: ["ent", "otolaryngology", "otolaryngologist", "ear nose throat"], symptoms: ["ear pain", "hearing loss", "sinus problem", "tonsil problem", "voice change"] },
  { code: "gynecology", name: "Gynecology", aliases: ["gynecology", "gynaecology", "gynecologist", "gynaecologist"], symptoms: ["menstrual problem", "pelvic pain", "cervical screening", "menopause", "ovarian cyst"] },
  { code: "obstetrics", name: "Obstetrics", aliases: ["obstetrics", "obstetrician", "maternity clinic", "prenatal clinic"], symptoms: ["pregnancy care", "prenatal visit", "high risk pregnancy", "postpartum review", "pregnancy complication"] },
  { code: "pediatrics", name: "Pediatrics", aliases: ["pediatrics", "paediatrics", "pediatrician", "paediatrician", "children's clinic"], symptoms: ["child fever", "growth concern", "child vaccination", "child illness", "developmental concern"] },
  { code: "psychiatry", name: "Psychiatry", aliases: ["psychiatry", "psychiatrist", "mental health clinic"], symptoms: ["depression", "anxiety", "psychosis", "mood changes", "sleep and mood problem"] },
  { code: "infectious-disease", name: "Infectious Disease", aliases: ["infectious disease", "infection specialist"], symptoms: ["persistent fever", "complex infection", "travel infection", "hiv care", "recurrent infection"] },
  { code: "allergy-immunology", name: "Allergy & Immunology", aliases: ["allergy and immunology", "allergy immunology", "allergist", "immunologist"], symptoms: ["seasonal allergy", "food allergy", "immune deficiency", "recurrent allergy", "hives"] },
  { code: "radiology", name: "Radiology", aliases: ["radiology", "radiologist", "diagnostic imaging"], symptoms: ["x-ray appointment", "ct scan", "mri scan", "ultrasound", "imaging-guided procedure"] },
  { code: "nuclear-medicine", name: "Nuclear Medicine", aliases: ["nuclear medicine", "pet scan", "radionuclide imaging"], symptoms: ["pet scan appointment", "thyroid uptake scan", "bone scan", "radionuclide therapy", "nuclear imaging"] },
  { code: "pathology", name: "Pathology", aliases: ["pathology", "pathologist", "laboratory pathology"], symptoms: ["biopsy review", "tissue diagnosis", "pathology second opinion", "laboratory tissue test", "cytology review"] },
  { code: "anesthesiology", name: "Anesthesiology", aliases: ["anesthesiology", "anaesthesiology", "anesthesiologist", "anaesthesiologist"], symptoms: ["preoperative anesthesia review", "sedation planning", "anesthesia complication review", "airway assessment", "perioperative assessment"] },
  { code: "pain-medicine", name: "Pain Medicine", aliases: ["pain medicine", "pain clinic", "pain specialist"], symptoms: ["chronic pain", "nerve pain", "complex regional pain", "back pain procedure", "cancer pain"] },
  { code: "rehabilitation-medicine", name: "Rehabilitation Medicine", aliases: ["rehabilitation medicine", "physical medicine and rehabilitation", "physiatry"], symptoms: ["stroke recovery", "injury rehabilitation", "mobility problem", "functional impairment", "prosthetic rehabilitation"] },
  { code: "sports-medicine", name: "Sports Medicine", aliases: ["sports medicine", "sports physician"], symptoms: ["sports injury", "overuse injury", "return to sport", "exercise-related pain", "concussion follow-up"] },
  { code: "critical-care-medicine", name: "Critical Care Medicine", aliases: ["critical care medicine", "intensive care", "icu", "intensivist"], symptoms: ["organ support", "severe illness", "ventilator care", "post intensive care review", "critical illness recovery"] },
  { code: "geriatrics", name: "Geriatrics", aliases: ["geriatrics", "geriatrician", "older adult clinic"], symptoms: ["frailty", "falls", "memory decline", "multiple medications", "older adult assessment"] },
  { code: "palliative-care", name: "Palliative Care", aliases: ["palliative care", "supportive care"], symptoms: ["advanced illness support", "symptom support", "goals of care", "caregiver support", "serious illness planning"] },
  { code: "medical-genetics", name: "Medical Genetics", aliases: ["medical genetics", "clinical genetics", "geneticist", "genetic counseling"], symptoms: ["inherited condition", "family history of genetic disease", "genetic test counseling", "congenital condition", "rare disease evaluation"] },
  { code: "preventive-medicine", name: "Preventive Medicine", aliases: ["preventive medicine", "preventative medicine", "preventive health"], symptoms: ["health screening", "vaccination planning", "risk assessment", "lifestyle prevention", "occupational health screening"] },
];

const FIRST_NAMES = [
  "Aarav", "Aisha", "Amelia", "Daniel", "Elena", "Elias", "Fatima", "Gabriel", "Hana",
  "Ibrahim", "Julia", "Kenji", "Leila", "Mateo", "Nadia", "Oliver", "Priya", "Samuel",
];
const LAST_NAMES = ["Adler", "Bennett", "Chen", "Dubois", "Garcia", "Hassan", "Ivanov", "Johansson"];

function generatedDoctor(index: number) {
  const first = FIRST_NAMES[index % FIRST_NAMES.length];
  const last = LAST_NAMES[Math.floor(index / FIRST_NAMES.length) % LAST_NAMES.length];
  return `Dr. ${first} ${last}`;
}

export const HOSPITAL_DEPARTMENTS: HospitalDepartment[] = DEPARTMENT_SEEDS.map((department, departmentIndex) => {
  const doctors = [0, 1, 2].map((slot) => generatedDoctor(departmentIndex * 3 + slot));
  return { ...department, doctors };
});

const allDoctorNames = HOSPITAL_DEPARTMENTS.flatMap((department) => department.doctors);
if (
  HOSPITAL_DEPARTMENTS.some((department) => department.doctors.length !== 3) ||
  new Set(allDoctorNames).size !== allDoctorNames.length
) {
  throw new Error("Hospital catalog requires exactly three globally unique doctors per department.");
}

export function departmentByCode(code: string) {
  return HOSPITAL_DEPARTMENTS.find((department) => department.code === code);
}

File: backend/app/hospital_catalog.py
import re


SPECIALTIES = [
    ("general-medicine", "General Medicine (Internal Medicine)", "persistent fever, fatigue, unexplained weight loss, multiple symptoms, adult checkup"),
    ("family-medicine", "Family Medicine", "routine checkup, cough, minor illness, preventive care, ongoing primary care"),
    ("emergency-medicine", "Emergency Medicine", "severe injury, chest pain, difficulty breathing, loss of consciousness, severe bleeding"),
    ("cardiology", "Cardiology", "palpitations, chest discomfort, high blood pressure, leg swelling, abnormal ECG"),
    ("neurology", "Neurology", "headache, seizure, numbness, tremor, memory problems"),
    ("neurosurgery", "Neurosurgery", "brain or spine surgery consultation, nerve compression, brain tumor surgery"),
    ("orthopedic-surgery", "Orthopedic Surgery", "joint pain, leg pain, fracture, back pain, sports injury"),
    ("general-surgery", "General Surgery", "hernia, gallbladder problem, appendix problem, abdominal lump, surgical wound"),
    ("plastic-surgery", "Plastic Surgery", "burn reconstruction, scar revision, hand reconstruction, complex wound"),
    ("vascular-surgery", "Vascular Surgery", "varicose veins, poor circulation, leg ulcer, arterial blockage"),
    ("thoracic-surgery", "Thoracic Surgery", "lung surgery consultation, chest wall mass, mediastinal mass, pleural disease"),
    ("cardiac-surgery", "Cardiac Surgery", "bypass surgery, heart valve surgery, aortic surgery, post-surgery review"),
    ("gastroenterology", "Gastroenterology", "abdominal pain, acid reflux, diarrhea, constipation, difficulty swallowing"),
    ("hepatology", "Hepatology", "jaundice, hepatitis, fatty liver, cirrhosis, abnormal liver tests"),
    ("pulmonology", "Pulmonology", "chronic cough, wheezing, shortness of breath, sleep apnea, abnormal lung test"),
    ("nephrology", "Nephrology", "chronic kidney disease, abnormal creatinine, protein in urine, dialysis care"),
    ("urology", "Urology", "urinary difficulty, kidney stones, prostate problem, blood in urine"),
    ("endocrinology", "Endocrinology", "diabetes, thyroid problem, hormone disorder, adrenal or metabolic disorder"),
    ("rheumatology", "Rheumatology", "joint stiffness, autoimmune disease, lupus, inflammatory arthritis, vasculitis"),
    ("hematology", "Hematology", "anemia, bleeding or clotting disorder, abnormal blood count, bone marrow problem"),
    ("oncology", "Oncology", "cancer evaluation, chemotherapy consultation, tumor follow-up, cancer survivorship"),
    ("dermatology", "Dermatology", "skin rash, acne, changing mole, skin lesion, eczema"),
    ("ophthalmology", "Ophthalmology", "vision changes, eye pain, cataract, glaucoma, retina problem"),
    ("ent", "ENT (Otolaryngology)", "ear pain, hearing loss, sinus problem, tonsil problem, voice change"),
    ("gynecology", "Gynecology", "menstrual problem, pelvic pain, cervical screening, menopause, ovarian cyst"),
    ("obstetrics", "Obstetrics", "pregnancy care, prenatal visit, high-risk pregnancy, postpartum review"),
    ("pediatrics", "Pediatrics", "child fever, growth concern, vaccination, child illness, developmental concern"),
    ("psychiatry", "Psychiatry", "depression, anxiety, psychosis, mood changes, sleep and mood problem"),
    ("infectious-disease", "Infectious Disease", "persistent fever, complex infection, travel infection, HIV care"),
    ("allergy-immunology", "Allergy & Immunology", "seasonal or food allergy, immune deficiency, recurrent allergy, hives"),
    ("radiology", "Radiology", "X-ray appointment, CT scan, MRI scan, ultrasound, imaging-guided procedure"),
    ("nuclear-medicine", "Nuclear Medicine", "PET scan, thyroid uptake scan, bone scan, radionuclide therapy"),
    ("pathology", "Pathology", "biopsy review, tissue diagnosis, pathology second opinion, cytology review"),
    ("anesthesiology", "Anesthesiology", "preoperative anesthesia review, sedation planning, airway assessment"),
    ("pain-medicine", "Pain Medicine", "chronic pain, nerve pain, complex regional pain, back pain procedure"),
    ("rehabilitation-medicine", "Rehabilitation Medicine", "stroke recovery, injury rehabilitation, mobility or functional impairment"),
    ("sports-medicine", "Sports Medicine", "sports injury, overuse injury, return to sport, exercise-related pain"),
    ("critical-care-medicine", "Critical Care Medicine", "organ support, severe illness, ventilator care, critical illness recovery"),
    ("geriatrics", "Geriatrics", "frailty, falls, memory decline, multiple medications, older adult assessment"),
    ("palliative-care", "Palliative Care", "advanced illness support, symptom support, goals of care, caregiver support"),
    ("medical-genetics", "Medical Genetics", "inherited condition, genetic family history, genetic test counseling, rare disease"),
    ("preventive-medicine", "Preventive Medicine", "health screening, vaccination planning, risk assessment, prevention"),
]

FIRST_NAMES = [
    "Aarav", "Aisha", "Amelia", "Daniel", "Elena", "Elias", "Fatima", "Gabriel", "Hana",
    "Ibrahim", "Julia", "Kenji", "Leila", "Mateo", "Nadia", "Oliver", "Priya", "Samuel",
]
LAST_NAMES = ["Adler", "Bennett", "Chen", "Dubois", "Garcia", "Hassan", "Ivanov", "Johansson"]

RAG_CORPUS_VERSION = "2026-07-28.2"
RAG_EMBEDDING_MODEL = "agentcare-private-semantic-hash-v1"

TERMINOLOGY = {
    "pain": ["pain", "pains", "painful", "paining", "hurts", "hurt", "aching", "burning", "stinging", "sore"],
    "urination": ["urination", "urinating", "urinate", "urine", "pee", "peeing", "pass urine", "passing urine", "pass water", "passing water"],
    "difficulty": ["difficulty", "difficult", "hard to", "unable to", "cannot", "can't", "trouble"],
    "frequency": ["frequent", "frequently", "often", "many times", "all the time"],
    "skin": ["skin", "body surface"],
    "rash": ["rash", "rashes", "spots", "itchy patches", "skin breakout"],
    "eye": ["eye", "eyes", "vision", "sight"],
    "ear": ["ear", "ears"],
    "hearing_loss": ["hearing loss", "cannot hear", "can't hear", "reduced hearing", "hearing problem"],
    "breathing": ["breathing", "breathe", "breath", "short of breath", "breathless"],
    "wheeze": ["wheeze", "wheezing", "whistling breath"],
    "abdomen": ["abdomen", "abdominal", "belly", "stomach", "tummy"],
    "reflux": ["acid reflux", "heartburn", "acid coming up", "sour taste"],
    "joint": ["joint", "joints", "knee", "knees", "shoulder", "shoulders", "hip", "hips"],
    "leg": ["leg", "legs", "limb", "limbs"],
    "fracture": ["fracture", "fractured", "broken bone", "bone break"],
    "heart_rhythm": ["palpitation", "palpitations", "heart racing", "racing heart", "heart pounding", "irregular heartbeat"],
    "heart_area": ["heart", "heart area", "cardiac area", "chest", "chest area", "centre of chest", "center of chest"],
    "pressure": ["pressure", "tightness", "tight", "heaviness", "squeezing"],
    "headache": ["headache", "head pain", "migraine"],
    "numbness": ["numbness", "numb", "pins and needles", "tingling"],
    "hives": ["hives", "welts", "allergic rash"],
    "thyroid": ["thyroid", "thyroid gland"],
    "menstrual": ["menstrual", "period problem", "periods", "menstruation"],
    "pregnancy": ["pregnancy", "pregnant", "prenatal", "expecting a baby"],
    "child": ["child", "children", "kid", "kids", "baby", "infant"],
    "fever": ["fever", "feverish", "high temperature"],
}

ROUTING_CONCEPTS = [
    ("urology-painful-urination", "urology", "painful urination", ["urination", "pain"], "route"),
    ("urology-voiding-difficulty", "urology", "difficulty passing urine", ["urination", "difficulty"], "route"),
    ("urology-frequency", "urology", "urinary frequency", ["urination", "frequency"], "route"),
    ("dermatology-skin-rash", "dermatology", "skin rash", ["skin", "rash"], "route"),
    ("ophthalmology-eye-pain", "ophthalmology", "eye pain", ["eye", "pain"], "review"),
    ("ent-ear-pain", "ent", "ear pain", ["ear", "pain"], "route"),
    ("ent-hearing-loss", "ent", "hearing problem", ["ear", "hearing_loss"], "review"),
    ("pulmonology-wheeze", "pulmonology", "wheezing", ["breathing", "wheeze"], "review"),
    ("gastroenterology-abdominal-pain", "gastroenterology", "abdominal pain", ["abdomen", "pain"], "review"),
    ("gastroenterology-reflux", "gastroenterology", "acid reflux", ["reflux"], "route"),
    ("orthopedics-joint-pain", "orthopedic-surgery", "joint pain", ["joint", "pain"], "review"),
    ("orthopedics-leg-pain", "orthopedic-surgery", "leg pain", ["leg", "pain"], "review"),
    ("orthopedics-fracture", "orthopedic-surgery", "fracture follow-up", ["fracture"], "route"),
    ("cardiology-heart-area-pain", "cardiology", "heart or chest-area pain", ["heart_area", "pain"], "review"),
    ("cardiology-heart-area-pressure", "cardiology", "heart or chest-area pressure", ["heart_area", "pressure"], "review"),
    ("cardiology-palpitations", "cardiology", "palpitations", ["heart_rhythm"], "review"),
    ("neurology-headache", "neurology", "headache", ["headache"], "review"),
    ("neurology-numbness", "neurology", "numbness or tingling", ["numbness"], "review"),
    ("allergy-hives", "allergy-immunology", "hives", ["hives"], "review"),
    ("endocrinology-thyroid", "endocrinology", "thyroid problem", ["thyroid"], "route"),
    ("gynecology-menstrual", "gynecology", "menstrual problem", ["menstrual"], "route"),
    ("obstetrics-pregnancy", "obstetrics", "pregnancy care", ["pregnancy"], "route"),
    ("pediatrics-child-fever", "pediatrics", "child fever", ["child", "fever"], "review"),
]


def doctor_name(index: int) -> str:
    return f"Dr. {FIRST_NAMES[index % len(FIRST_NAMES)]} {LAST_NAMES[(index // len(FIRST_NAMES)) % len(LAST_NAMES)]}"


def validate_catalog() -> None:
    names = [doctor_name(index) for index in range(len(SPECIALTIES) * 3)]
    if len(SPECIALTIES) != 42 or len(names) != len(set(names)):
        raise RuntimeError("Hospital catalog must contain 42 specialties and globally unique doctors.")


def canonical_terms(value: str) -> set[str]:
    text = value.lower()
    return {
        canonical
        for canonical, synonyms in TERMINOLOGY.items()
        if any(
            re.search(
                r"\b" + re.escape(synonym).replace(r"\ ", r"\s+") + r"\b",
                text,
            )
            for synonym in synonyms
        )
    }


def cardiovascular_safety_signal(value: str) -> bool:
    terms = canonical_terms(value)
    return "heart_area" in terms and bool({"pain", "pressure"} & terms)


def concepts_for_department(department_code: str):
    return [concept for concept in ROUTING_CONCEPTS if concept[1] == department_code]


def match_routing_concepts(value: str):
    terms = canonical_terms(value)
    matches = [
        concept
        for concept in ROUTING_CONCEPTS
        if all(required in terms for required in concept[3])
    ]
    scores: dict[str, int] = {}
    for _, department_code, _, required, _ in matches:
        scores[department_code] = scores.get(department_code, 0) + len(required)
    ranked = sorted(scores.items(), key=lambda item: (-item[1], item[0]))
    leading = matches[0] if matches else None
    unique_lead = bool(
        leading
        and ranked
        and ranked[0][0] == leading[1]
        and (len(ranked) == 1 or ranked[0][1] > ranked[1][1])
    )
    return {
        "canonical_terms": sorted(terms),
        "matches": matches,
        "leading": leading,
        "can_route": bool(leading and leading[4] == "route" and unique_lead),
    }
